### ISSUE TITLE
imfile module to read existing logs

### ISSUE THREAD

**Issue by ven000** [rsyslog/rsyslog#5422]
### Expected behavior
read existing logs
### Actual behavior
only read real time logs
### Steps to reproduce the behavior
module(load="imfile" PollingInterval="10")  # Load the imfile module
input(type="imfile"
      File="/remotelogs/*/*.log"
      tag="historical-log"
      Severity="info"
      Facility="local7")
### Environment
- rsyslog version: 8.2112.0 (aka 2021.12)
- platform: Ubuntu 16.04
- for configuration questions/issues, include rsyslog.conf and included config files

# /etc/rsyslog.conf configuration file for rsyslog
#
# For more information install rsyslog-doc and see
# /usr/share/doc/rsyslog-doc/html/configuration/index.html
#
# Default logging rules can be found in /etc/rsyslog.d/50-default.conf

#################
#################

module(load="imuxsock") # provides support for local system logging

module(load="imfile" PollingInterval="10")  # Load the imfile module
input(type="imfile"
      File="/remotelogs/*/*.log"
      Tag="historical-log"
      Severity="info"
      Facility="local7")

#module(load="immark")  # provides --MARK-- message capability

# provides UDP syslog reception
module(load="imudp")
input(type="imudp" port="514")

# provides TCP syslog reception
module(load="imtcp")
input(type="imtcp" port="514")

$template RemInputLogs, "/remotelogs/%FROMHOST-IP%/%PROGRAMNAME%.log"
*.* ?RemInputLogs

# provides kernel logging support and enable non-kernel klog messages
module(load="imklog" permitnonkernelfacility="on")

###########################
###########################

#
# Use traditional timestamp format.
# To enable high precision timestamps, comment out the following line.
#
$ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat

# Filter duplicated messages
$RepeatedMsgReduction on

#
# Set the default permissions for all log files.
#
$FileOwner syslog
$FileGroup adm
$FileCreateMode 0640
$DirCreateMode 0755
$Umask 0022
$PrivDropToUser syslog
$PrivDropToGroup syslog

#
# Where to place spool and state files
#
$WorkDirectory /var/spool/rsyslog

#
# Include all config files in /etc/rsyslog.d/
#
$IncludeConfig /etc/rsyslog.d/*.conf
*.* @@ip

**Comment by davidelang** [comment]
On Tue, 30 Jul 2024, ven000 wrote:

> ### Expected behavior
> read existing logs
> ### Actual behavior
> only read real time logs

by this do you mean that it's only reading new logs that are written, not all 
the logs in the file?

according to the info you included below, you are running a very old version of 
rsyslog (from 2021), imfile has been re-written since then.

you can try setting freshstarttail to false in case your old version defaults to 
it being on, but you really need to update to the current version.

https://www.rsyslog.com/doc/configuration/modules/imfile.html#freshstarttail

David Lang

> ### Steps to reproduce the behavior
> module(load="imfile" PollingInterval="10")  # Load the imfile module
> input(type="imfile"
>      File="/remotelogs/*/*.log"
>      tag="historical-log"
>      Severity="info"
>      Facility="local7")
> ### Environment
> - rsyslog version: 8.2112.0 (aka 2021.12)
> - platform: Ubuntu 16.04
> - for configuration questions/issues, include rsyslog.conf and included config files
>
> # /etc/rsyslog.conf configuration file for rsyslog
> #
> # For more information install rsyslog-doc and see
> # /usr/share/doc/rsyslog-doc/html/configuration/index.html
> #
> # Default logging rules can be found in /etc/rsyslog.d/50-default.conf
>
>
> #################
> #### MODULES ####
> #################
>
> module(load="imuxsock") # provides support for local system logging
>
> module(load="imfile" PollingInterval="10")  # Load the imfile module
> input(type="imfile"
>      File="/remotelogs/*/*.log"
>      Tag="historical-log"
>      Severity="info"
>      Facility="local7")
>
> #module(load="immark")  # provides --MARK-- message capability
>
> # provides UDP syslog reception
> module(load="imudp")
> input(type="imudp" port="514")
>
> # provides TCP syslog reception
> module(load="imtcp")
> input(type="imtcp" port="514")
>
> $template RemInputLogs, "/remotelogs/%FROMHOST-IP%/%PROGRAMNAME%.log"
> *.* ?RemInputLogs
>
> # provides kernel logging support and enable non-kernel klog messages
> module(load="imklog" permitnonkernelfacility="on")
>
>
> ###########################
> #### GLOBAL DIRECTIVES ####
> ###########################
>
> #
> # Use traditional timestamp format.
> # To enable high precision timestamps, comment out the following line.
> #
> $ActionFileDefaultTemplate RSYSLOG_TraditionalFileFormat
>
> # Filter duplicated messages
> $RepeatedMsgReduction on
>
> #
> # Set the default permissions for all log files.
> #
> $FileOwner syslog
> $FileGroup adm
> $FileCreateMode 0640
> $DirCreateMode 0755
> $Umask 0022
> $PrivDropToUser syslog
> $PrivDropToGroup syslog
>
> #
> # Where to place spool and state files
> #
> $WorkDirectory /var/spool/rsyslog
>
> #
> # Include all config files in /etc/rsyslog.d/
> #
> $IncludeConfig /etc/rsyslog.d/*.conf
> *.* @@ip
>
> 
>
>
>

**Comment by ven000** [comment]
yes, it worked once reading the existing old logs, at the same time, it generate a tag.log file in the remotelogs folder, which has existing log data. Previously I didn't know why a new log file with tag name generated, then I delete that tag.log file. after that, it stopped reading the existing log files, and it always generate a tag named log file, tag.log, but that file doesn't have existing log data anymore, and i couldn't see the existing/old logs in kibana.

**Comment by davidelang** [comment]
ven000 wrote:

> yes, it worked once reading the existing old logs, at the same time, it 
> generate a tag.log file in the remotelogs folder, which has existing log data. 
> Previously I didn't know why a new log file with tag name generated, then I 
> delete that tag.log file. after that, it stopped reading the existing log 
> files, and it always generate a tag named log file, tag.log, but that file 
> doesn't have existing log data anymore, and i couldn't see the existing/old 
> logs in kibana.

are you really running the version from 2021? if so, please update to a current 
version.

it would also be useful to get pstats output from your system. You have things 
in the config that could block and cause it to stop processing logs as queues 
and network buffers fill up.

are all logs stopping? or are you only having problems with logs from imfile? 
There were some versions several years ago that had problems with inotify, so 
you could try switching to polling. But upgrading to a newer version that has 
the re-write of imfile would be far better than spending a lot of time 
troubleshooting such an old version (assuming you are really running one from 
2021)

David Lang

**Comment by ven000** [comment]
> ven000 wrote:
> yes, it worked once reading the existing old logs, at the same time, it generate a tag.log file in the remotelogs folder, which has existing log data. Previously I didn't know why a new log file with tag name generated, then I delete that tag.log file. after that, it stopped reading the existing log files, and it always generate a tag named log file, tag.log, but that file doesn't have existing log data anymore, and i couldn't see the existing/old logs in kibana.
> are you really running the version from 2021? if so, please update to a current version. it would also be useful to get pstats output from your system. You have things in the config that could block and cause it to stop processing logs as queues and network buffers fill up. are all logs stopping? or are you only having problems with logs from imfile? There were some versions several years ago that had problems with inotify, so you could try switching to polling. But upgrading to a newer version that has the re-write of imfile would be far better than spending a lot of time troubleshooting such an old version (assuming you are really running one from 2021) David Lang

yes, the other guy installed the rsyslog and elk server. would the elk still work if using the latest version of rsyslog? that's my concern

**Comment by davidelang** [comment]
yes, rsyslog is very careful to maintain backwards compatibility

David Lang
